src/base/rss/rss_autodownloader.cpp

```diff
@@ -0,0 +1,379 @@
+    {
+        Logger::instance()->addMessage(QString("Couldn't save RSS AutoDownloader data in %1. Error: %2")
+                                       .arg(fileName).arg(errorString), Log::WARNING);
+    });
+
+    m_ioThread->start();
+
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::downloadFromUrlFinished
+            , this, &AutoDownloader::handleTorrentDownloadFinished);
+    connect(BitTorrent::Session::instance(), &BitTorrent::Session::downloadFromUrlFailed
+            , this, &AutoDownloader::handleTorrentDownloadFailed);
+
+    load();
+
+    m_processingTimer->setSingleShot(true);
+    connect(m_processingTimer, &QTimer::timeout, this, &AutoDownloader::process);
+
+    if (m_processingEnabled)
+        startProcessing();
+}
+
+AutoDownloader::~AutoDownloader()
+{
+    store();
+
+    m_ioThread->quit();
+    m_ioThread->wait();
+}
+
+AutoDownloader *AutoDownloader::instance()
+{
+    return m_instance;
+}
+
+bool AutoDownloader::hasRule(const QString &ruleName) const
+{
+    return m_rules.contains(ruleName);
+}
+
+AutoDownloadRule AutoDownloader::ruleByName(const QString &ruleName) const
+{
+    return m_rules.value(ruleName, AutoDownloadRule("Unknown Rule"));
+}
+
+QList<AutoDownloadRule> AutoDownloader::rules() const
+{
+    return m_rules.values();
+}
+
+void AutoDownloader::setRule(const AutoDownloadRule &rule)
+{
+    if (!m_rules.contains(rule.name())) {
+        // Insert new rule
+        setRule_impl(rule);
+        emit ruleAdded(rule.name());
+        resetProcessingQueue();
+        m_dirty = true;
+    }
+    else if (m_rules[rule.name()] != rule) {
+        // Update existing rule
+        setRule_impl(rule);
+        emit ruleChanged(rule.name());
+        resetProcessingQueue();
+        m_dirty = true;
+    }
+
+    store();
+}
+
+bool AutoDownloader::renameRule(const QString &ruleName, const QString &newRuleName)
+{
+    if (!m_rules.contains(ruleName)) return false;
+    if (m_rules.contains(newRuleName)) return false;
+
+    m_rules.insert(newRuleName, m_rules.take(ruleName));
+    m_dirty = true;
+    emit ruleRenamed(newRuleName, ruleName);
+    store();
+    return true;
+}
+
+void AutoDownloader::removeRule(const QString &ruleName)
+{
+    if (m_rules.remove(ruleName) > 0) {
+        m_dirty = true;
+        emit ruleRemoved(ruleName);
+    }
+
+    store();
+}
+
+void AutoDownloader::process()
+{
+    if (m_processingQueue.isEmpty()) return; // processing was disabled
+
+    processJob(m_processingQueue.takeFirst());
+    if (!m_processingQueue.isEmpty())
+        // Schedule to process the next torrent (if any)
+        m_processingTimer->start();
+
+    store();
+}
+
+void AutoDownloader::handleTorrentDownloadFinished(const QString &url)
+{
+    auto job = m_waitingJobs.take(url);
+
+    auto feed = Session::instance()->feedByURL(job->feedURL);
+    if (feed) {
+        if (auto article = feed->articleByGUID(job->articleGUID))
```

Comment:
Well, it's a matter of taste... Maybe I'll rework it in places where it will not break readability.